### PR TITLE
Add Keycloak Authorization dynamic tenant config resolution

### LINKED ISSUE #17664: Support the dynamic KeycloakPolicyEnforcerAuthorizer tenants
## Description

KeycloakPolicyEnforcerAuthorizer tenants configured in `application.properties` are supported in 2.0.0.x but it is not possible to create them programmatically yet as can be done for OIDC tenants with `TenantConfigResolver`.

## Implementation ideas

Add `TenantPolicyConfigResolver` which would accept `RoutingContext` and the current OIDC `tenantId` and return `KeycloakPolicyEnforcerTenantConfig` and have it injected in `PolicyEnforcerResolver`. 

`PolicyEnforcerResolver` would check `TenantPolicyConfigResolver` first - if it returns `KeycloakPolicyEnforcerTenantConfig` then  `Function` will be used to convert it to `PolicyEnforcer` - this function will be passed from `KeycloakRecorder` to `PolicyEnforcerResolver` in its constructor similarly to how it is done in `OidcRecorder`.

It should be enforced for `web-app` tenants only (as is already done for the configured `web-app` tenants) that the current `OidcTenantConfig` also has `quarkus.oidc.roles.source=accesstoken`. The current dynamic `OidcTenantConfig` is available as `RoutingContext` `dynamic.tenant.config` attribute.

A simple test would have to be added to `integration-tests/keycloak-authorization` - it already has one test for the configured non-default tenant - so another similar test should be added for the dynamically created one

CC @pedroigor

### LINKED ISSUE #25774: Dynamic Keycloak Policy Enforcer for multi-tenancy
### Description

Support creating keycloak authorization policy dynamically

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
- Closes: #25774
- Closes: #17664

### CODE REVIEW

**Comment by michalvavrik** [comment]
hey @pedroigor please review if you would like, otherwise in terms of what is changed, changes are not specific to Keycloak (it's a CDI feature...), so @sberyozkin is right reviewer for this.

File: docs/src/main/asciidoc/security-keycloak-authorization.adoc [hunk 1]
```diff
@@ -559,6 +559,61 @@ quarkus.keycloak.webapp-tenant.policy-enforcer.paths.1.paths=/api/permission
...
+
+@ApplicationScoped
+public class CustomTenantPolicyConfigResolver implements TenantPolicyConfigResolver {
+
+    private final KeycloakPolicyEnforcerTenantConfig enhancedTenantConfig;
+    private final KeycloakPolicyEnforcerTenantConfig newTenantConfig;
+
+    public CustomTenantPolicyConfigResolver(KeycloakPolicyEnforcerConfig enforcerConfig) {
+        KeycloakPolicyEnforcerTenantConfig tenantConfig = enforcerConfig.defaultTenant();
+        this.enhancedTenantConfig = KeycloakPolicyEnforcerTenantConfig.builder(tenantConfig) <1>
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: Builder is interesting, can we try to simplify it somehow ? There are 3 build() calls for a fairly simple policy, `get` is mentioned twice.
IMHO it is worth trying to make as compact as possible.
I'd consider dropping `add`, use concrete verb method and do `.path("enhanced-config-path").get()` or `method(GETenum)`  etc. IMHO we should have a single `build` terminating the whole sequence only.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: I will look into this. We certainly can simplify it. Single `build` I'm not sure, it would have to be `buildAll` and other parent builder aware `builds`. But things like `get()` certainly can be done. I didn't think about it as my only objection was to make it possible to configure what you can with app properties. Will try to improve it.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: I don't think we can remove what is there now as right now, it's one to one to documentation and to configuration in config properties. That makes it more verbose but clear what you need to set. I have user experience with extensive io.fabric8 k8 builders and it's hard to find right method and right path to the end.

However we can provide shortcuts for simple scenarios and document them. I'll have a look again.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: My point is, it makes sense to create "clever" methods for builders that you use often, but for builder like this one, main objection must be to have straightforward way to determine what to call.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: I added few shortucs where I though it makes sense. I don't think it makes sense to optimize this too much because you don't want to spend time studying what some method does. It's easier to have it verbous. But please feel free to suggest improvements.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: added build all option as well

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: @michalvavrik Hi, the way this builder is created will serve as a blueprint for how OIDC and OIDC client configurations can be created after `new OidcTenantConfig()` and `new OidcClientConfig()` are deprecated. However, `new OidcTenantConfig()` and `new OidcClientConfig()` options are very straightforward and easy for users to use, which is why it is a quite a big deal IMHO how this builder API is designed. 

If `buildAll()` works, why have individual build() ?

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: > @michalvavrik Hi, the way this builder is created will serve as a blueprint for how OIDC and OIDC client configurations can be created after `new OidcTenantConfig()` and `new OidcClientConfig()` are deprecated. 

So glad you consider this! Migration rules :-)

> However, `new OidcTenantConfig()` and `new OidcClientConfig()` options are very straightforward and easy for users to use, which is why it is a quite a big deal IMHO how this builder API is designed.

I'm in no hurry, keep asking for changes.
 
> If `buildAll()` works, why have individual `build()` ?

You have tenant config builder, and then you have nested builders. Nested builders can close all the builders with `buildAll` and build tenant config. However nested builders can also want to keep continue, like add another path, or cache or some other setting. We can't avoid nested `build`s methods or `buildAll`. Or do you have alternative proposal? Sorry, I realize you asked for something else, but I tried to explain how annoying it can be too try to figure what build method should I use. That's why this builder has also available methods that are one to one to ConfigMapping interface.

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: >However nested builders can also want to keep continue, like add another path, or cache or some other setting.

IMHO this is unnecessary, I believe the focus should be on letting users do a very compact configuration creation, as opposed to keeping the flexibility like the one you described: one adds a path group, then does something else, then return to that path group to add something else. If the builder API offers an option to configure some composite set of properties, users can take that option if they need to and set whatever they need in the composite option, there is no need to keep multiple composite groups open

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: That sets:

```
quarkus.keycloak.policy-enforcer.path-cache.max-entries
quarkus.keycloak.policy-enforcer.path-cache.lifespan
```

You can see that without any knowledge or thinking what  `.pathCache().maxEntries(5).lifespan(2).build()` does, I don't need to look for methods or think what they smartly do.What is alternative for this without builder?

I think this:

`.pathCache().maxEntries(5).pathCache().lifespan(2)` or maybe `pathCache().maxEntriedAndLifespan(5, 2)`

Is that what you mean? Every developer knows builder patter, I think what is there is easier, but this is hardly important for me. Just confirm @sberyozkin and there will be only one `build()` please, thank you

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: Sorry, the start of previous comment disappeared :/,here it is:

 in the tests below you can see `.pathCache().maxEntries(5).lifespan(2).build()`

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: Alright, I'll do that.

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:591]: @michalvavrik We can experiment with a few options, may be `set` is unnecessary as you have `pathCache()` without any parameters so it is quite distinctive.
We can look at how a builder sequence looks where all possible properties are set :-) and see what can be optimized

**Comment by michalvavrik** [comment]
Done, the `DynamicTenantConfigPolicyEnforcerTest` tests new builder comprehensively.

**Comment by sberyozkin** [comment]
Hi @michalvavrik, can you please type a  builder sequence showing all the supported properties being set, I was not sure about `setGet("write")`...

**Comment by michalvavrik** [comment]
> Hi @michalvavrik, can you please type a builder sequence showing all the supported properties being set

Please have a look at the `io.quarkus.it.keycloak.DynamicTenantConfigPolicyEnforcerTest#assertEveryConfigPropertyCanBeSet` added in this PR.

> I was not sure about `setGet("write")`...

It's varargs, so you can ignore scope if you want, ATM I am not throwing an exception when there is no scope because I wasn't able to determine if there can be such a scenario where it could be valid to avoid scope but set method. Anyway, according to the https://quarkus.io/guides/all-config#quarkus-keycloak-authorization_quarkus-keycloak-policy-enforcer-paths-paths-methods-methods-scopes it is required when the method is specified, and you wanted shortcuts (for POST, GET, PUT, ...) and want it compact. Not sure how to do both? Please feel free to provide suggestion.

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        config = KeycloakPolicyEnforcerTenantConfig.builder().setPaths("/x", "/y").setEnforcementMode(PERMISSIVE).build();
+        pathConfig = config.policyEnforcer().paths().entrySet().stream().findAny().get().getValue();
+        assertEquals(2, pathConfig.paths().get().size());
+        assertTrue(pathConfig.paths().get().contains("/x"));
+        assertTrue(pathConfig.paths().get().contains("/y"));
+        assertEquals(PERMISSIVE, pathConfig.enforcementMode());
+    }
+
+    private static void assertEveryConfigPropertyCanBeSet() {
+        var builder = KeycloakPolicyEnforcerTenantConfig.builder()
```
Comment (sberyozkin) [inline integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: @michalvavrik It looks fine, but indeed, my earlier suggestion to use `set` does not look great, instead,
```
 KeycloakPolicyEnforcerTenantConfig.builder()
                .enforcementMode(DISABLED)
                .claimInformationPoint(Map.of("one", Map.of("two", "three")), Map.of("four", Map.of()))
                .connectionPoolSize(-1)
                .lazyLoadPaths(false)
                .pathCache(5, 2)
                .httpMethodAsScope(true)
                .build();
```
would be nicer, sorry for suggesting adding `set`.

Can you also please give me a favor and add one more test, showing the same properties but which will expand shortcuts for the groups like `.pathCache(5, 2)` into `.pathCache().prop1(5).prop2(2)`.

I'm sorry it is a bit time consuming, but IMHO the more compact the sequence the better the user experience will be, as we agreed, they would provide a template for the OidcTenantConfig and OidcClientConfig. Thanks for your patience :-)

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+            assertEquals("TRACE", method.method());
+            // default scope enforcement is ALL
+            assertEquals(PolicyEnforcerConfig.ScopeEnforcementMode.ALL, method.scopesEnforcementMode());
+        });
+        // create config from previous config and alter one of paths
+        var builder = KeycloakPolicyEnforcerTenantConfig.builder(config);
+        var path = builder.setPaths("/api/one");
+        path.setPermissionName("New Permission");
+        path.setClaimInformationPoint(Map.of("key", Map.of("sub-key", "sub-value")));
+        path.clearMethods();
```
Comment (sberyozkin) [inline integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:110]: I'm not sure it is necessary. Users will set up the config and build it, in a client scope, where they can reuse the same builder instance for creating multiple configurations, having clear methods might be useful, but not in the server scope IMHO. `SecurityIdentity` builder does not have clear methods for example

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    private static void assertBuilderPopulatedWithDefaultValues() {
+        var config = KeycloakPolicyEnforcerTenantConfig.builder().build();
+        assertEquals(20, config.connectionPoolSize());
+        assertTrue(config.policyEnforcer().lazyLoadPaths());
+        assertEquals(ENFORCING, config.policyEnforcer().enforcementMode());
+        assertTrue(config.policyEnforcer().paths().isEmpty());
+        assertFalse(config.policyEnforcer().httpMethodAsScope());
+        assertEquals(30000, config.policyEnforcer().pathCache().lifespan());
+        assertEquals(1000, config.policyEnforcer().pathCache().maxEntries());
+        assertTrue(config.policyEnforcer().claimInformationPoint().simpleConfig().isEmpty());
```
Comment (sberyozkin) [inline integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:83]: `simpleConfig` and `complexConfig` should not be part of the Config builder API

Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:83]: it is not, here the `config` is an interface annotated with `@ConfigMapping`. These lines assert that build config builder contains expected values.

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java (hunk 3 above)
Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:110]: I added it for very edge case, in case you just want to adjust original configuration (remove same methods without rewriting them). I'll drop it. It's no biggie.

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java (hunk 2 above)
Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: np, on it

> Can you also please give me a favor and add one more test, showing the same properties but which will expand shortcuts for the groups like .pathCache(5, 2) into .pathCache().prop1(5).prop2(2).

well, that is no longer possible. I leveraged the fact that maxEntries and lifespan has different data type so you can do either `pathCache(5,2)` or `pathCache(5)` or `pathCache(2)`. You can read method parameter names and I can add a javadoc (though TBH I think the generated javadoc by IDE is more then sufficient). IMO it is not confusing because whatever you do, you need to read method parameters in order to know what you are setting. Personally I don't know almost any methods by heart.

Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: image

Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: image

Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: I've added `DynamicTenantConfigPolicyEnforcerTest#assertPathCacheConfigOnly` that shows all the path cache combinations.

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java (hunk 3 above)
Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java]: Done.

Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java]: Removed.

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java (hunk 4 above)
Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:84]: > `simpleConfig` and `complexConfig` should not be part of the Config builder API

they are not, so closing this comment thread as resolved

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java (hunk 2 above)
Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: > It looks fine, but indeed, my earlier suggestion to use set does not look great, instead,

I've dropped `set` and added path cache tests that address questions previous. I'll close this comment thread as i believe it will be easier to discuss possible new questions separately.

Comment (sberyozkin) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: @michalvavrik It looks nice, but losing `.pathCache().a().b()` and the same for other groups is not great IMHO.

Do you mean that with `.pathCache().a().b()` we have to close it with `build()` ?

So either

`KeycloakPolicyBuilder.builder()
  .pathCache(1, 2)
  .build()`

or 
```
KeycloakPolicyBuilder.builder()
  .pathCache().a(1).b(2).build()
  .build()
```
?

See if the group has say 7 or more properties, a few of such groups are available in OIDC, then users would have to type `builder.myGroup(1, null, 3, null) as it is nearly impossible to offer all possible overrides for users to choose the most suitable one where they can avoid setting `null`.

So not sure we can really avoid supporting itin addition to the shortcuts like `.pathCache(1, 2)`

May be a tiny optimization is possible in this case:

```
KeycloakPolicyBuilder
  .pathCache().a(1).b(2).build()
  .someOtherGroup().c(1).d(2).e().build()
  .build()
```

where `KeycloakPolicyBuilder.pathCache()` inits the main builder and starts the specific group builder so the users do not have to type  `.builder()`.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * A tenant resolver is responsible for resolving the {@link KeycloakPolicyEnforcerTenantConfig} for tenants, dynamically.
+ */
+public interface TenantPolicyConfigResolver {
+
+    /**
+     * Returns a {@link KeycloakPolicyEnforcerTenantConfig} given a {@code RoutingContext} and tenant id.
+     *
+     * @param routingContext the routing context
+     * @param tenantId tenant id
```
Comment (sberyozkin) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:18]: I was think earlier, how this resolver aligns with the OIDC tenant resolution logic. Is this tenantId representing an OIDC tenant id resolved at the OIDC level ?

Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:18]: Yep, I didn't change logic behind where this id comes from (just moved it around). This is current main:

https://github.com/quarkusio/quarkus/blob/243f46c10c4ac0cb380af120d942a9559a075b82/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerAuthorizer.java#L63

Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:18]: _I think_ (not sure) I only called it `tenantId` because in current main, it is kinda shared tenant, you select it `quarkus.keycloak."tenant".connection-pool-size` so it is OIDC tenant, but it is also Keycloak tenant. If there is such a thing. I suppose there is at least theoretically.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+     * @param requestContext request context
+     *
+     * @return the tenant configuration. If the uni resolves to {@code null}, indicates that the default
+     *         configuration/tenant should be chosen
+     */
+    Uni<KeycloakPolicyEnforcerTenantConfig> resolve(RoutingContext routingContext, String tenantId,
+            KeycloakRequestContext requestContext);
+
+    /**
+     * Keycloak Context that can be used to run blocking tasks.
```
[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:28]: pfff :-D good question, but I simply thought it would be weird to use `OidcRequestContext` inside Keycloak extension. It's different place. If you are sure that `OidcRequestContext` will also be relevant for Kecloak and you want add there non-Keycloak stuff in the future, let's reuse it. I'd love to drop another class, but I am not sure it is a smart move in regards of future breaking changes.

File: integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java (hunk 2 above)
Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: > Do you mean that with .pathCache().a().b() we have to close it with build() ?

there is no such thing like `pathCache().a().b()`, only configmapping has it, we have 3 setters: `pathCache(1,2)`, `pathCache(1L)`, `pathCache(2)`.

> See if the group has say 7 or more properties, a few of such groups are available in OIDC, then users would have to type builder.myGroup(1, null, 3, null) as it is nearly impossible to offer all possible overrides for users to choose the most suitable one where they can avoid setting null.

currently there is only one group like that, it is `Path` and for that, I created it's own builder because I don't like `builder.myGroup(1, null, 3, null)`. so it works like this:

```
var pathBuilder = builder.namedPaths("p1", "path");
pathBuilder.enforcementMode(PolicyEnforcerConfig.EnforcementMode.PERMISSIVE);
pathBuilder.permissionName("n1");
pathBuilder.method("method1", PolicyEnforcerConfig.ScopeEnforcementMode.DISABLED, "scope1", "scopes2");
var config = builder.build();
```

What I like about this is that there is no `build`, there is no `saving` or `end()`, it is build as soon as you build the parent builder and then, it builds immutable configuration interface. You can keep using the original builder without any risk.

> So not sure we can really avoid supporting itin addition to the shortcuts like .pathCache(1, 2)

As long as we can settle on something, it's absolutely fine to create these builders again for users...

> May be a tiny optimization is possible in this case:

That's not a tiny optimization, it will require me to create all of these sub-builders again. I thought you wanted to drop them. Your example has 2 `build()` again, which I though you wrote somewhere is not necessary. I am happy to do it if you are certain @sberyozkin .

I don't see appeal in doing `pathCache().a(1).b(2).build().some...` over `pathCache(1, 2)` but it is important to get it right, so if you are absolutely certain @sberyozkin , let's do it.

Comment (sberyozkin) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: Hi @michalvavrik 

I did not mean `pathCache().a(1).b(2).build()` was better than `pathCache(1, 2)` but the former is easier to understand by just looking at the existing code, what `a` and `b` means, `pathCache(1, 2)` is clear enough when we type it in IDE, but not afterwards.

> currently there is only one group like that, it is Path and for that, I created it's own builder because I don't like builder.myGroup(1, null, 3, null). so it works like this:
```
 var pathBuilder = builder.namedPaths("p1", "path");
pathBuilder.enforcementMode(PolicyEnforcerConfig.EnforcementMode.PERMISSIVE);
pathBuilder.permissionName("n1");
pathBuilder.method("method1", PolicyEnforcerConfig.ScopeEnforcementMode.DISABLED, "scope1", "scopes2");
var config = builder.build();
```

Sure, so here you close everything with the parent's builder's `build()`, but users can't do it all in a single sequence so a builder specific `build()` would only be needed for the single sequence expression. 

IMHO, it should be possible to build every group in one of several ways and it should be consistent for every group:
* builder.group1(1, 2,3,4,5).group2(5,6,7,8,9).build(); // and add as many reasonable overloads for a given group as needed
* builder.group1().a1(1).b1(1).end().group2().a2(2).b2(2).end().build(); 

I'm using `end()` following your comment above to indicate a group specific end - users won't have to use it - it is only there to support a single sequence. But as you showed, users would then also be able to skip them with a single `build()` if they want to skip group specific builds. 

>That's not a tiny optimization, it will require me to create all of these sub-builders again. 

There I was only referring to the option where you skip initial `.builder()` - but if you agree to the earlier suggestions, you would be creating individual group builders in any case...

I agree it all may look  like a bit of over-engineering, but IMHO it is worth trying to make it as optimal and consistent as we can, like we agreed, it is a template for providing future OIDC builders

Comment (michalvavrik) [reply integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/DynamicTenantConfigPolicyEnforcerTest.java:182]: ok, I'll do it.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java (hunk 5 above)
Comment (sberyozkin) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:18]: I was thinking about it, IMHO it makes sense to pass `OidcTenantConfig` representing the current tenant since here it is a custom code, what will it do with this tenant id alone, not clear. FYI, DefaultTenantConfigResolver has a helper like  `getResolvedTenant(String tenantId)`

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:28]: I'll use `OidcRequestContext` then.

[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:28]: Thanks

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java (hunk 5 above)
Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/TenantPolicyConfigResolver.java:18]: I think your ideas are good. I know what they would do with the tenant id (identity tenant exactly as we do, via id so that we know which config belongs to him), but having whole config is much more comfortable for the user.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/PolicyEnforcerResolver.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * A {@link PolicyEnforcer} resolver.
+ */
+public interface PolicyEnforcerResolver {
+
+    Uni<PolicyEnforcer> resolvePolicyEnforcer(RoutingContext routingContext, String tenantId);
+
+    long getReadTimeout();
```
Comment (sberyozkin) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/PolicyEnforcerResolver.java:15]: Why this method should be here ?

Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/PolicyEnforcerResolver.java:15]: It is tenant specific https://quarkus.io/guides/all-config#quarkus-keycloak-authorization_quarkus-keycloak-tenant-connection-pool-size and if someone decides to define it's own `PolicyEnforcerResolver` than this timeout is used when `resolvePolicyEnforcer` is invoked, so you might need longer read timeout for certain tenants. That is in the configuration already.

We provide our `PolicyEnforcerResolver` for users and while in CDI world we can't stop users from defining their own, like #25774 asked for, users shouldn't have reasons to do that.

Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/PolicyEnforcerResolver.java:15]: Sorry, wrong link, let me update the comment in the minute.

**Comment by michalvavrik** [comment]
I've rebased on the current main and resolved merge conflicts just in case @sberyozkin find a time to review.

**Comment by michalvavrik** [comment]
@sberyozkin please have a look; I won't resolve merge conflicts again as I'm bit worried it's wasting of CI resources. Will do it when you find a time for this

**Review by sberyozkin — APPROVED** [review]
Thanks @michalvavrik , let's do it and start planning doing something similar for OIDC 🙂
Cheers Sergey
